#! /bin/bash

#SBATCH -N 1
#SBATCH -n 1
#SBATCH -c 1
#SBATCH --mem=62000
#SBATCH --time=48:00:00
#SBATCH --partition=largemem
#SBATCH --account=berglandlab


### Filtering of the VCF generated by VarScan. Last modified 3/25/2020


### LOADING MODULES 
	echo "loading modules"
	module load gcc/7.1.0
	module load vcftools/0.1.15
	module load bedtools/2.26.0


### Define some variables
	echo "defining directories and such"
	inputDir="/scratch/ab5dr/multiSpecies/deNovoCalls"
	outDir="/scratch/ab5dr/multiSpecies/vcfs"
	scriptDir="/scratch/ab5dr/multiSpecies/scripts"


### select the reference genome and repmasker files
   species=${1}
   if [ ${species} = "Dmel" ]; then
      ref="/scratch/ab5dr/pooledSeq2019/poolSeqPipe/misc/holo_dmel_6.12.fa"
      rep="/scratch/ab5dr/multiSpecies/ref/repeats/Dmelv6.repmasker.bed"
   elif [ ${species} = "Dsim" ]; then
      ref="/scratch/ab5dr/multiSpecies/ref/GCF_016746395.1_Prin_Dsim_3.0_genomic.fna"
      rep="/scratch/ab5dr/multiSpecies/ref/repeats/Dsim.repmasker.edit.bed"
   elif [ ${species} = "Dhyd" ]; then
      ref="/scratch/ab5dr/multiSpecies/ref/GCF_003285905.1_DhydRS2_genomic.fna"
      rep="/scratch/ab5dr/multiSpecies/ref/repeats/Dhyd.repmasker.edit.bed"
   elif [ ${species} = "Zind" ]; then  
      ref="/scratch/ab5dr/multiSpecies/ref/z_indianus_16GNV01_v02.fasta"
      rep="/scratch/ab5dr/multiSpecies/ref/repeats/z_indianus_16GNV01_v02.putative_function.drop.sort.repeat.bed"
   elif [ ${species} = "Dsuz" ]; then
      ref="/scratch/ab5dr/multiSpecies/ref/GCF_013340165.1_LBDM_Dsuz_2.1.pri_genomic.fna"
      rep="/scratch/ab5dr/multiSpecies/ref/repeats/Dsuz.repmasker.edit.bed"
   else
      echo "error"
   fi



### Build +/- 10 bp file to remove SNPs near indels
	echo "building indel bed"
	cat ${inputDir}/MS_${species}.VarScan.indel.vcf | sed -n -e '/^#CHROM/,$p' | cut -f1,2 | awk -v pad=10 '{print $1"\t"$2-pad"\t"$2+pad}' > ${scriptDir}/MS_${species}_indel_10bp.bed


### Remove regions around indels
	echo "removing regions around indels"
	bedtools intersect -v -header \
	-a ${inputDir}/MS_${species}.VarScan.newHeader.vcf \
	-b ${scriptDir}/MS_${species}_indel_10bp.bed > \
	${outDir}/MS_${species}.VarScan.newHeader.noIndelReg.vcf


### Remove repetitive regions
	echo "filtering out repetitive regions"
	bedtools intersect -v -header \
	-a ${outDir}/MS_${species}.VarScan.newHeader.noIndelReg.vcf \
	-b ${rep} > \
	${outDir}/MS_${species}.VarScan.newHeader.noIndelReg.noRep.vcf


### Remove non-biallelic sites 
	echo "filtering out non-biallelic sites"
	vcftools --vcf ${outDir}/MS_${species}.VarScan.newHeader.noIndelReg.noRep.vcf --min-alleles 2 --max-alleles 2 --recode --recode-INFO-all --out ${outDir}/MS_${species}.VarScan.newHeader.noIndelReg.noRep.biAllelic



